core/src/main/java/org/apache/calcite/plan/RelOptUtil.java

```diff
@@ -3304,6 +3304,29 @@ public static RelNode createProject(RelNode child, Mappings.TargetMapping mappin
     return createProject(projectFactory, child, Mappings.asListNonNull(mapping.inverse()));
   }
 
+  /** Returns the relational table node for {@code tableName} if it occurs within a
+   * relational expression {@code root} otherwise an empty option is returned. */
+  public static @Nullable RelOptTable findTable(RelNode root, final String tableName) {
+    RelOptTable[] table = new RelOptTable[1];
+    try {
+      RelShuttle visitor = new RelHomogeneousShuttle() {
+        @Override public RelNode visit(TableScan scan) {
+          final RelOptTable scanTable = scan.getTable();
+          final List<String> qualifiedName = scanTable.getQualifiedName();
+          if (qualifiedName.get(qualifiedName.size() - 1).equals(tableName)) {
+            table[0] = scanTable;
+            throw Util.FoundOne.NULL;
```

Comment:
a bit simpler is `throw new Util.FoundOne(scanTable)`; then you don't need the `table[]` variable.